// ftpd/index.js
// YellowLeaf FTP by Michiel Dral 
var EventEmitter, FtpClient, FtpDataPool, FtpServer, Promise, SuperClass, debug, fs, ftpserver, net, path,
  __slice = [].slice,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

net = require("net");

path = require("path");

Promise = require('bluebird');

fs = Promise.promisifyAll(require("fs"));

EventEmitter = require("events").EventEmitter;

SuperClass = require("superclass");

debug = require('debug')('[FTP]', 'cyan');

FtpDataPool = require('./ftpdatapool');

EventEmitter.prototype.emitWith = function() {
  var args, event, scope;
  event = arguments[0], scope = arguments[1], args = 3 <= arguments.length ? __slice.call(arguments, 2) : [];
  return this.listeners(event).forEach(function(fn) {
    return fn.apply(scope, args);
  });
};

FtpServer = (function(_super) {
  __extends(FtpServer, _super);

  function FtpServer(connectionListener) {
    connectionListener = typeof connectionListener === "function" ? connectionListener : function() {};
    this.server = net.createServer((function(_this) {
      return function(socket) {
        return _this.emitWith('connection', new FtpClient(socket));
      };
    })(this));
    this.on('connection', connectionListener);
    ['address'].forEach((function(_this) {
      return function(method) {
        return _this[method] = _this.server[method];
      };
    })(this));
  }

  ['listen', 'close'].forEach(function(method) {
    return FtpServer.prototype[method] = function() {
      var args, _ref;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      (_ref = this.server)[method].apply(_ref, args);
      return this;
    };
  });

  FtpServer.prototype.debug = function(pleasedo) {
    if (pleasedo) {
      return debug = require('debug')('[FTP]', 'cyan');
    } else {
      return debug = function() {};
    }
  };

  return FtpServer;

})(EventEmitter);

FtpClient = (function(_super) {
  __extends(FtpClient, _super);

  function FtpClient(connection) {
    this.host = connection.localAddress;
    this._connection = connection;
    this.cwd = '/';
    this.basedir = process.cwd();
    this.dataServer = new FtpDataPool(this);
    this._connection.on('readable', (function(_this) {
      return function() {
        var args, cmd, data, options, _ref;
        data = _this.read();
        if (!data) {
          return;
        }
        _ref = data.split(' '), cmd = _ref[0], args = 2 <= _ref.length ? __slice.call(_ref, 1) : [];
        cmd = cmd.toLowerCase();
        options = [];
        args = args.filter(function(arg) {
          if (arg.indexOf('-') === 0) {
            options.push(args);
            return false;
          } else {
            return true;
          }
        });
        cmd.options = options;
        debug("Command: " + cmd + " with " + args);
        if (_this.listeners('command.' + cmd).length < 1) {
          _this.emit('unknownCommand', cmd, args.join(' '));
        }
        return _this.emit.call(_this, 'command.' + cmd, args.join(' '));
      };
    })(this));
    this.write('220 Welcome!');
  }

  FtpClient.prototype.getFullPath = function(file) {
    if (file == null) {
      file = '';
    }
    if (this.cwd == null) {
      this.cwd = '/';
    }
    if (this.basedir == null) {
      this.basedir = '/';
    }
    if (!file.startsWith('/')) {
      file = path.join(this.cwd, file);
    }
    return path.join(this.basedir, file);
  };


  /*
  fs: (method, file, args...) ->
    currpath = @getFullPath file
    fs[method + "Async"](currpath, args...)
    .catch (err) =>
      @write "550 Can't fly here, this place does not exist!"
      err.ftpNotified = true
      throw err
   */

  FtpClient.prototype.write = function(data) {
    return this._connection.write(data + '\r\n');
  };

  FtpClient.prototype.read = function(size) {
    var data;
    data = this._connection.read(size);
    if (!data) {
      return;
    }
    return data.toString().trim();
  };

  FtpClient.prototype.end = function() {
    this._connection.end();
    return this.dataServer = void 0;
  };

  return FtpClient;

})(EventEmitter);

module.exports = FtpServer;

module.exports.FtpClient = FtpClient;

module.exports.FtpDataPool = FtpDataPool;

module.exports.defaults = {
  dataSocket: function() {
    this.on('command.pasv', function() {
      return this.dataServer.openPasv();
    });
    return this.on('command.port', function(args) {
      return this.dataServer.openPort(args);
    });
  },
  unknownCommand: function() {
    return this.on('unknownCommand', function(cmd, args) {
      debug("Command " + cmd + " not implemented (args: " + args + ").");
      return this.write('500 Sorry.');
    });
  }
};

if (!module.parent) {
  ftpserver = new FtpServer;
  ftpserver.listen(800);
}
